libbeat/common/dtfmt/elems.go

```diff
@@ -50,6 +50,14 @@ type paddedNumber struct {
 	signed               bool
 }
 
+type paddedBigNumber struct {
```

Comment:
@urso, i add a new elements. It works, what do you think?

Reply:
I don't think we need this structure if we introduce (or extend) the opcodes to also stores a `div` up to 32bit. `paddedNumber` would need to choose the right opcode based on the size of div. The we wouldn't need to cast to `float` or use math.Pow.

Reply:
If we wants to stores a div up to 32bit, i think we should change `prog` structure, is it?

Reply:
I don't think we need to change the structure. For example see opCopyLong (it uses a 16bit length counter). Putting a 32bit value into prog can easily be done via `binary.LittleEndian.PutUint32`.

Checking all usages of paddedNumber we could actually change all uses of paddedNumber to report the exponent for 10^exp (like you did for divLen). 

```
type paddedNumber struct {
  ft fieldType
  divExp int // <- replace div with exponent for computing 10^exp
  minDigits, maxDigits int
  signed bool
}

func (n paddedNumber) compile() (prog, error) {
  if n.divExp == 0 {
    ...
  }
  return makeProg(opExtNumPadded, byte(n.ft), byte(n.divExp), byte(n.maxDigits))
}
```

And use a table instead of math.Pow.
This changes smeantics of opExtNumPadded, but I didn't find any place where we need the actual divisor, so that's fine:

```
var pow10Table [10]int

func init() {
  x := 1
  for i := range pow10Table {
    pow10Table[i] = x
    x *= 10
  }
}

...
		case opExtNumPadded:
			ft, divExp, digits := fieldType(p.p[i]), int(p.p[i+1]), int(p.p[i+2])
            div := pow10Table[divExp]
			i += 3
			v, err := getIntField(ft, ctx, t)
			if err != nil {
				return bytes, err
			}
			bytes = appendPadded(bytes, v/div, digits)
```

What's nice about this is that we don't need to add any new cases to prog. We can also remove opNumPadded now, because a `divExp` value of 0 will divide by 1 anyways.

I think this later approach is more close to the implementation you provided, but also unifies all fraction handling in the package to use the pow10 based approach of yours (minus the float operations).

Reply:
If we remove `div` and add `divExp` , the `paddednumber` only support 10^exp, and i'm not sure whether it affect the `appendDecimalValue` function,  so i add a `paddedBigNumber` structure before.

I'm thinking change the `compile()` function like below when `div` too large, it use `binary.PutVarint` save the int to byte and `binary.Varint` to convert byte to int now. But i'm not sure this way whether correct.

This way maybe not good, what do you think?

```go
func (n paddedNumber) compile() (prog, error) {
	if n.div == 0 {
		return makeProg(opNumPadded, byte(n.ft), byte(n.maxDigits))
	}
        if n.div > 255 {
                buf := make([]byte, 4)
                bvar := binary.PutVarint(buf, int64(div))
                bufLen := len(buf[:n])
                newBuf := make([]byte, 0, bufLen)
                newBuf = append(newBuf, opExtNumPadded, byte(n.ft), byte(bufLen)))
                newBuf = append(newBuf, bvar ...)
                newBuf = append(newBuf, byte(n.maxDigits))
                return makeProg(newBuf...)
       }
        
	return makeProg(opExtNumPadded, byte(n.ft), byte(n.div), byte(n.maxDigits))
}
```

Reply:
> If we remove div and add divExp , the paddednumber only support 10^exp, and i'm not sure whether it affect the appendDecimalValue function, so i add a paddedBigNumber structure before.

The `appendDecimalValue` method always sets `div` to 0. The only call that sets the `div` is `appendExtDecimal`. All uses of `appendExtDecimal` have a divisor that is dividable by 10. This is why I did propose this approach.

Your approach with PutVarint will also work. Using Varint will add some decoding overhead whenever we want to print the data. this is why I opted for using a 32bit value by using `binary.LittleEndian.PutUint32`.